usr/man/_vax.d/man3/ctype.3
.\" SCCSID: @(#)ctype.3	2.7	8/13/87
.TH ctype 3 
.SH Name
isalpha, isupper, islower, isdigit, isxdigit, isalnum, isspace, ispunct, isprint, isgraph, iscntrl, isascii \- character classification macros
.SH Syntax
.B #include <ctype.h>
.PP
.B int isalpha (\fIc\fP)
.PP
.B int c;
.SH Description
.NXR "ctype keyword"
.NXR "isalpha subroutine"
.NXR "isupper subroutine"
.NXR "islower subroutine"
.NXR "isdigit subroutine"
.NXR "isalnum subroutine"
.NXR "isspace subroutine"
.NXR "ispunct subroutine"
.NXR "isprint subroutine"
.NXR "isprint subroutine"
.NXR "isgraph subroutine"
.NXR "iscntrl subroutine"
.NXR "isascii subroutine"
.NXR "ASCII character" "classifying"
.PP
These macros classify character-coded integer values according to the rules of the coded character set (codeset) identified by the last successful call to
.PN setlocale
category
.PN LC_CTYPE .
All macros return non-zero for true and zero for false.
.PP
If
.PN setlocale
category
.PN LC_CTYPE
has not been called successfully, or if character classification information is not available for a supported language, then characters are classified according to the rules of the ASCII 7-bit coded character set, returning 0 for values above octal 0177.
.PP
The macro
.PN isascii 
provides a result for all integer values.
The rest provide a result for EOF and values in the character range of the codeset identified by the last successful call to
.PN setlocale
category
.PN LC_CTYPE .
.TP 15n
.PN isalpha
.I c
is a letter
.TP
.PN isupper
.I c
is an uppercase letter
.TP
.PN islower 
.I c
is a lowercase letter
.TP
.PN isdigit
.I c
is a digit
.TP
.PN isxdigit
.I c
is a hexadecimal digit, by default [0-9], [A-F], or [a-f]
.TP
.PN isalnum
.I c
is an alphanumeric character
.TP
.PN isspace
.I c
is a space, tab, carriage return, new line, or form feed
.TP
.PN ispunct
.I c
is a punctuation character (neither control, alphanumeric, nor space)
.TP
.PN isprint
.I c
is a printing character, by default code 040(8) (space) through 0176 (tilde)
.TP
.PN isgraph
.I c
is a printing character, like
.PN isprint
except false for space
.TP
.PN iscntrl
.I c
is a delete character (0177) or ordinary control character
(less than 040) except for space characters
.TP
.PN isascii
.I c
is an ASCII character, code less than 0200
.SS International Environment
.IP LC_CTYPE 15
If this environment variable is set and valid,
.PN ctype
uses the international language database named in the definition to 
determine character classification rules.
.IP LANG
If this environment variable is set and valid, 
.PN ctype
uses the international language database named in the definition to determine
the character classification rules. If 
.PN LC_CTYPE
is defined, that definition supercedes the definition of 
.PN LANG .
.SH See Also
conv(3), setlocale(3), stdio(3s), environ(5int), ascii(7)
.br
.I "Guide to Developing International Software"


lib/libc/gen/ctype.c
#ifndef lint
static	char	*sccsid = "@(#)ctype.c	4.1	ULTRIX	7/3/90";
#endif lint

/************************************************************************
 *									*
 *			Copyright (c) 1985, 1988 by			*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/
/************************************************************************
 *			Modification History				*
 *									*
 *									*
 * 004  Dennis Doherty, Jan 15 1990					*
 *	Added new extern that is set to the size of the _pctype         *
 *	table. (_pctype_siz) This is used by the multi-byte to 		*
 *	Wide char functions.						*
 *									*
 * 003  Andy Gadsby,  27-Oct-1987					*
 *	Extended table to make each entry 2 bytes and to cater for	*
 *	8 bit characters. This is required for internationalization.	*
 *	To prevent old applications picking up the wrong table the name *
 *	has been modified by removing an underscore.			*
 *									*
 *	David L Ballenger, 11-Oct-1985					*
 * 002	Change definitions so that iscntrl(), ispunct(), and isgraph()	*
 *	behave as documented, and so that the array is the same size in *
 *	both the System V and ULTRIX environments.			*
 *									*
 *	David L Ballenger, 30-Mar-1985					*
 * 0001	Add defintions for System V compatibility.			*
 *									*
 ************************************************************************/

/* @(#)ctype_.c	4.2 (Berkeley) 7/8/83 */

#include	<ctype.h>

/* This array is used by the following macros: isalpha(), isupper(), 
 * islower(), isdigit(), isxdigit(), isalnum(), isspace(), ispunct(),
 * isprint(), isgraph(), iscntrl(), and provides defined results for
 * all integer values for which the macro isascii() is true (ie. 0-0177),
 * plus 0200-0377 for internationalization, and the non-ASCII value 
 * EOF (ie. -1).  All of the macros which use this array index off the 
 * base address + 1 to achieve this range, ie. _ctype__+1[c].
 */

unsigned short _ctype__[1 + 256] = {
	0,
	_C,	_C,	_C,	_C,	_C,	_C,	_C,	_C,
	_C,	_C|_S,	_C|_S,	_C|_S,	_C|_S,	_C|_S,	_C,	_C,
	_C,	_C,	_C,	_C,	_C,	_C,	_C,	_C,
	_C,	_C,	_C,	_C,	_C,	_C,	_C,	_C,
	_S|_B,	_P,	_P,	_P,	_P,	_P,	_P,	_P,
	_P,	_P,	_P,	_P,	_P,	_P,	_P,	_P,
	_N,	_N,	_N,	_N,	_N,	_N,	_N,	_N,
	_N,	_N,	_P,	_P,	_P,	_P,	_P,	_P,
	_P,	_U|_X,	_U|_X,	_U|_X,	_U|_X,	_U|_X,	_U|_X,	_U,
	_U,	_U,	_U,	_U,	_U,	_U,	_U,	_U,
	_U,	_U,	_U,	_U,	_U,	_U,	_U,	_U,
	_U,	_U,	_U,	_P,	_P,	_P,	_P,	_P,
	_P,	_L|_X,	_L|_X,	_L|_X,	_L|_X,	_L|_X,	_L|_X,	_L,
	_L,	_L,	_L,	_L,	_L,	_L,	_L,	_L,
	_L,	_L,	_L,	_L,	_L,	_L,	_L,	_L,
	_L,	_L,	_L,	_P,	_P,	_P,	_P,	_C,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0,
	 0,	 0,	 0,	 0,	 0,	 0,	 0,	 0
};

unsigned short *_pctype = _ctype__;	/* pointer to current table */

unsigned long _pctype_siz = 255;	/* max char code for this table */
